fix display of undefined on card name after page refresh

huult opened the pull request:
### Details

### Fixed Issues
$ https://github.com/Expensify/App/issues/52140
PROPOSAL: https://github.com/Expensify/App/issues/52140#issuecomment-2462343948

### Tests

Same QA step

- [x] Verify that no errors appear in the JS console

### Offline tests

### QA Steps

1. Open the Expensify app.
2. Create a workspace, click on More features, and enable Company cards.
3. Add a company card.
4. Click on the Settings button on the newly added company card.
5. Click Card feed name.
6. Refresh the page.
7. Verify that the card name is displayed correctly without showing "Undefined Card." on input.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/300231f5-f6b5-4dcf-a3ba-0edd53c51189

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/fd35eac7-946c-40af-b79d-2556852cb602

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/b8eada0e-6ae7-4ff5-b497-bd80fe059e13

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/fd058201-1cb3-4c5f-825c-ce3e1e9906df

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/ecc77fe8-5394-4af3-94e1-62f715d1b226

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/63759071-4433-416f-84ee-035b7fd02733

</details>

huult committed:
fix display of undefined on card name after page refresh

huult committed:
Add type parameter for company card setting feed name

DylanDylann commented:
@huult As this [comment](https://github.com/Expensify/App/issues/52140#issuecomment-2467143061), I mean that we pass feedName to route when navigating to Setting Page

huult committed:
pass selected feed name when navigating to Settings Page

huult commented:
@DylanDylann I've updated it. Sorry for missing that.

DylanDylann commented on src/pages/workspace/companyCards/WorkspaceCompanyCardsSettingsPage.tsx:
```diff
@@ -45,7 +45,7 @@ function WorkspaceCompanyCardsSettingsPage({
 
     const navigateToChangeFeedName = () => {
-        Navigation.navigate(ROUTES.WORKSPACE_COMPANY_CARDS_SETTINGS_FEED_NAME.getRoute(policyID));
+        Navigation.navigate(ROUTES.WORKSPACE_COMPANY_CARDS_SETTINGS_FEED_NAME.getRoute(policyID, feedName));
```
```suggestion
        Navigation.navigate(ROUTES.WORKSPACE_COMPANY_CARDS_SETTINGS_FEED_NAME.getRoute(policyID, selectedFeed));
```

DylanDylann commented on src/libs/Navigation/types.ts:
```diff
@@ -854,7 +854,7 @@ type SettingsNavigatorParamList = {
     [SCREENS.WORKSPACE.COMPANY_CARDS_SETTINGS_FEED_NAME]: {
         policyID: string;
-        currentFeedName?: string;
+        feedName?: string;
```
```suggestion
        selectedFeed: string;
```

DylanDylann replied:
Please update in other places

huult replied:
@DylanDylann We can't use `selectedFeed` because it will make things more complex. If you want to use `selectedFeed`, we must also send the `cardFeeds` object in the parameters.

huult replied:
If we use selectedFeed without passing the cardFeeds object, we encounter an issue like the video below because cardFeeds is retrieved from Onyx.
<details>
<summary>VIDEO</summary>
https://github.com/user-attachments/assets/ee547f38-cdd2-4de7-8517-11ee8f394c4b
</details>

DylanDylann replied:
In src/pages/workspace/companyCards/WorkspaceCompanyCardsSettingsFeedNamePage.tsx, we can get cardFeed from Onyx, right?

huult replied:
yes

DylanDylann replied:
Could you elaborate in more detail on what problem you want to mention [here](https://github.com/Expensify/App/pull/52301#discussion_r1836132308)?

huult replied:
I'm still not clear on your comment. Can you explain in more detail?

- [Here](https://github.com/Expensify/App/pull/52301#discussion_r1836102753), We're just renaming that, right?
- [Here](https://github.com/Expensify/App/pull/52301#discussion_r1836101691), Do you want me to send selectedFeed instead of feedName?

In my opinion, we can't use selectedFeed to set the parameters because we have a bug when using it, and I mentioned the reason in the comment. If we rename it, then I think we should rename feedName to selectedFeedName

DylanDylann replied:
Let's pass selectedFeed in the param and then we can handle your bug in another way. Please check [this branch](https://github.com/DylanDylann/App/tree/setting-feed-name-new-params) if there is any confusion. Thanks.

huult replied:
@DylanDylann I understand your point; you're suggesting we should use selectedFeed instead of selectedFeedName. However, if we use selectedFeed, we encounter the issue I showed in the video. Are you aware of this, and if so, do you have any suggestions for handling it? Because if we get the value from [cardFeeds](https://github.com/DylanDylann/App/blob/e61493bb25f87d71fb0a83f1b36a5d47392f25a4/src/pages/workspace/companyCards/WorkspaceCompanyCardsSettingsFeedNamePage.tsx#L45), we won’t need to set additional parameters.

DylanDylann replied:
small note: Never give a video without any description and say it is a bug like in the video. When elaborating on a bug please say detailed steps, actual behavior, and expected behavior or at least a description to other people can understand that you want to say

> we encounter an issue like the video below because cardFeeds is retrieved from Onyx.

DylanDylann replied:
>  Are you aware of this, and if so, do you have any suggestions for handling it?

It seems I handled it in my ref branch

DylanDylann replied:
> we won’t need to set additional parameters.

Also fine, if that we need to handle loading indicator if lastSelectedFeed isn't loaded

huult replied:
@DylanDylann , Can you test this case on your branch?
https://github.com/user-attachments/assets/ed1f1681-6527-469d-b97b-cf07a5eb8cea

huult replied:
> Also fine, if that we need to handle loading indicator if lastSelectedFeed isn't loaded

@DylanDylann I will update it to use selectedFeed to set the parameters and add a loading indicator to wait for the [cardFeeds](https://github.com/Expensify/App/blob/ae40ea5843e18a05a42222fe78a29aacf06c5b46/src/pages/workspace/companyCards/WorkspaceCompanyCardsSettingsFeedNamePage.tsx#L42) value. Do you agree?

huult replied:
Got it. Will update. thank you

DylanDylann replied:
> Can you test this case on your branch?

@huult It worked. Is there any problem?

https://github.com/user-attachments/assets/93387765-15bd-4ba5-b10f-9c532a7bac23

DylanDylann commented:
@huult Also please update your videos

DylanDylann commented:
Please add one more step to verify the route display the correct selected feed when clicking on Card feed name

huult commented:
> Please add one more step to verify the route display the correct selected feed when clicking on Card feed name
@DylanDylann I added that to step 6.

DylanDylann commented:
@huult 

> Also fine, if that we need to handle loading indicator if lastSelectedFeed isn't loaded

Could you try this approach if it works well we don't need to add param anymore

huult committed:
Merge remote-tracking branch 'upstream/main' into 52140-undefined-card-name-on-refresh

huult committed:
Add loading to load Onyx value

huult commented:
@DylanDylann We are able to use the loading indicator, which is my alternative solution that I just updated. I also updated the attachment step and code. Can you review it again?

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.
